crates/c-api/src/lib.rs

```diff
@@ -488,15 +489,34 @@ pub unsafe extern "C" fn wasm_func_call(
         let val = &(*args.add(i));
         params.push(val.val());
     }
-    match func.call(&params) {
-        Ok(out) => {
+
+    // We're calling arbitrary code here most of the time, and we in general
+    // want to try to insulate callers against bugs in wasmtime/wasi/etc if we
+    // can. As a result we catch panics here and transform them to traps to
+    // allow the caller to have any insulation possible against Rust panics.
+    let result = panic::catch_unwind(AssertUnwindSafe(|| func.call(&params)));
+    match result {
+        Ok(Ok(out)) => {
             for i in 0..func.result_arity() {
                 let val = &mut (*results.add(i));
                 *val = wasm_val_t::from_val(&out[i]);
             }
             ptr::null_mut()
         }
-        Err(trap) => {
+        Ok(Err(trap)) => {
+            let trap = Box::new(wasm_trap_t {
+                trap: HostRef::new(trap),
+            });
+            Box::into_raw(trap)
+        }
+        Err(panic) => {
+            let trap = if let Some(msg) = panic.downcast_ref::<String>() {
```

Comment:
Would it be possible to include the panic's backtrace in the trap?

Reply:
Resolved after discussion with Alex (short answer is: it's possible, long answer is: diagnostics can already be printed with `RUST_BACKTRACE` anyway).  Not needed at this time.

Reply:
Ah yeah unfortunately rust panics are just `Box<Any + Send>` so they don't have a backtrace associated with them. To capture a backtrace we'd have to install a panic hook which means dealing with weird global/thread local state/etc, so I figure it's best to hold out for this unless we need it.